resolver_netip.go

```diff
@@ -120,41 +151,28 @@ func eqNetIP(ip1, ip2 netip.Addr) bool {
 // revive:disable:cognitive-complexity
 func msgToNetIPq(m *dns.Msg, qType uint16) ([]netip.Addr, error) {
 	// revive:enable:cognitive-complexity
-	s := []netip.Addr{}
-	if !validMsg(m) {
-		return nil, errBadMessage
-	}
-	if dns.TypeToString[qType] == "A" {
-		for _, rec := range m.Answer {
-			if rec.Header().Rrtype == dns.TypeA {
-				if ip, ok := netip.AddrFromSlice(rec.(*dns.A).A); ok {
+	var s []netip.Addr
+
+	if m != nil && m.Rcode == dns.RcodeSuccess {
+		switch qType {
+		case dns.TypeA:
+			ForEachAnswer(m, func(r *dns.A) {
+				if ip, ok := netip.AddrFromSlice(r.A); ok {
 					s = append(s, ip)
 				}
-			}
-		}
-	}
-	if dns.TypeToString[qType] == "AAAA" {
-		for _, rec := range m.Answer {
-			if rec.Header().Rrtype == dns.TypeAAAA {
-				if ip, ok := netip.AddrFromSlice(rec.(*dns.AAAA).AAAA); ok {
+			})
+		case dns.TypeAAAA:
+			ForEachAnswer(m, func(r *dns.AAAA) {
+				if ip, ok := netip.AddrFromSlice(r.AAAA); ok {
 					s = append(s, ip)
 				}
-			}
+			})
 		}
-	}
-	return s, nil
-}
 
-func validMsg(m *dns.Msg) bool {
```

Comment:
Better than removing it, maybe a better approach is not move it centrally and make it callable for any `*dns.Msg`